[core] Reset zoom history state in still mode

brunoabinader opened the pull request:
```
$ node platform/node/test/render.test.js --recycle-map \
    "line-dasharray/slant" \
    "line-dasharray/literal--line-width-composite-function"
* passed line-dasharray slant
* failed line-dasharray literal--line-width-composite-function
1 passed (50.0%)
1 failed (50.0%)
```

| expected | actual | diff |
| -- | -- | -- |
| ![expected](https://user-images.githubusercontent.com/76133/29885386-441976aa-8dbf-11e7-96cf-3fb2bb6b6999.png) | ![actual](https://user-images.githubusercontent.com/76133/29885388-491ed74e-8dbf-11e7-8533-f469bcd97fa8.png) | ![diff](https://user-images.githubusercontent.com/76133/29885396-500b38fe-8dbf-11e7-911a-fe9b160a92a4.png) |

This is an edge case from the issue reported in https://github.com/mapbox/mapbox-gl-native/pull/9760. When switching from a higher zoom level to a lower one, we might end up evaluating render layers for _zoom + 1_ which causes the blurred lines.

jfirebaugh commented:
I don't quite understand the issue. What's the sequence of zooms that triggers it and why does this fix it?

brunoabinader commented:
> I don't quite understand the issue. What's the sequence of zooms that triggers it and why does this fix it?

OK, so first off there are some assumptions to keep in mind:
- We use `ZoomHistory.lastIntegerZoom` to evaluate cross-faded paint properties (such as `line-dasharray`).
- The bool [zoomChanged](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/renderer/renderer_impl.cpp#L75) is used, among other checks, to check whether to [re-evaluate render layers](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/renderer/renderer_impl.cpp#L166).

Now let's check the workflow below - lines starting with `[current]` describes the ZoomHistory state prior to [zoomHistory.update](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/renderer/renderer_impl.cpp#L75) call, and `[updated]` after.

> * passed line-dasharray slant
[Renderer::Impl::render]
    -- [current] lastZoom: 18.00 lastFloorZoom: 18.00 lastIntegerZoom: 18.00
    -- [updated] lastZoom: 0.50 lastFloorZoom: 0.00 lastIntegerZoom: 1.00 **zoomChanged: true**
    -- [layer: road] added: false changed: true zoomChanged: true hasTransition: false
        *** **re-evaluating render layer road** ***
[Renderer::Impl::render]
    -- [current] lastZoom: 0.50 lastFloorZoom: 0.00 lastIntegerZoom: 1.00
    -- [updated] lastZoom: 0.50 lastFloorZoom: 0.00 **lastIntegerZoom: 0.00** **zoomChanged: false**
    -- [layer: road] added: false changed: false zoomChanged: false hasTransition: false
> * failed line-dasharray literal--line-width-composite-function

As we can see from the first `Renderer::Impl::render` call, the current `lastZoom` left from the previous test run was `18.0`. With the new one being lower - `0.5` - we [end up](https://github.com/mapbox/mapbox-gl-native/blob/master/src/mbgl/map/zoom_history.hpp#L30) with `lastIntegerZoom` being `1.0`. Zoom has changed, so we re-evaluate the `road` layer accordingly.

Now, on the second `render` call, `lastIntegerZoom` value changes because it [differs from the floored zoom value](https://github.com/mapbox/mapbox-gl-native/blob/0694d224c745ea5ee5eb1fc46cbb0b9487e6333a/src/mbgl/map/zoom_history.hpp#L32) (we added this in #9760 to fix a similar issue). However, because neither zoom nor the layer itself actually changes - possibly due to many different places triggering `Map::Impl::onUpdate` - we don't re-evaluate the layer. So the `line-dasharray` property evaluation ends up using a now-outdated value, which causes the blurred lines :boom:.

**The change I propose in this PR is to cause `zoomChanged` to be true also when `lastIntegerZoom` value changes - which then forces the render layers to be re-evaluated, causing the cross-faded paint properties to evaluate using the now-updated `lastIntegerZoom` value.**

---- 

There are also a few other things to consider:
- Evaluating cross-faded paint properties using an _underzoomed_ integer zoom level leads to rendering artifacts (such as the blurred lines in `line-dasharray` seen above). I wonder if there is any change in our shader code to prevent that.

- Also, this makes me believe the layer re-evaluation pass from the first render call is potentially bogus and redundant. We could also prevent that by adding something like this:
```diff
diff --git a/src/mbgl/renderer/renderer_impl.cpp b/src/mbgl/renderer/renderer_impl.cpp
index 21ff8ed..7f36c57 100644
--- a/src/mbgl/renderer/renderer_impl.cpp
+++ b/src/mbgl/renderer/renderer_impl.cpp
@@ -72,6 +72,10 @@ void Renderer::Impl::render(const UpdateParameters& updateParameters) {
     
     assert(BackendScope::exists());
 
+    if (updateParameters.mode == MapMode::Still) {
+        zoomHistory.first = true;
+    }
+
     const bool zoomChanged = zoomHistory.update(updateParameters.transformState.getZoom(), updateParameters.timePoint);
 
     const TransitionParameters transitionParameters {
```

jfirebaugh commented:
> * passed line-dasharray slant
[Renderer::Impl::render]
    -- [current] lastZoom: 18.00 lastFloorZoom: 18.00 lastIntegerZoom: 18.00
    -- [updated] lastZoom: 0.50 lastFloorZoom: 0.00 lastIntegerZoom: 1.00 **zoomChanged: true**
    -- [layer: road] added: false changed: true zoomChanged: true hasTransition: false
        *** **re-evaluating render layer road** ***
[Renderer::Impl::render]
    -- [current] lastZoom: 0.50 lastFloorZoom: 0.00 lastIntegerZoom: 1.00
    -- [updated] lastZoom: 0.50 lastFloorZoom: 0.00 **lastIntegerZoom: 0.00** **zoomChanged: false**
    -- [layer: road] added: false changed: false zoomChanged: false hasTransition: false
> * failed line-dasharray literal--line-width-composite-function

The unexpected part in the sequence to me is the fact that `lastIntegerZoom` changes from 1 to 0 even though there was no change in the zoom level. That shouldn't happen -- I think it's an unintended result from #9760.

brunoabinader commented:
> The unexpected part in the sequence to me is the fact that lastIntegerZoom changes from 1 to 0 even though there was no change in the zoom level. That shouldn't happen -- I think it's an unintended result from #9760.

Given these impediments, I don't see a better way besides:

1. Partially revert #9760: remove the ```|| lastIntegerZoom != floorZ``` check
2. Set `zoomHistory.first` to true if map is in still mode

I don't think we should be using zoomHistory's zoom state management for still mode: I suspect some rounding problem is happening in shader code due to float usage - which causes the rendering artifacts. To avoid that, by resetting zoomHistory state we'd assure zoom is always bigger than ```lastIntegerZoom```.

What do you think? I'm going to update this PR accordingly.

jfirebaugh commented:
@ansis, can you step in here and advise?
